### PR TITLE
Add tax fields for distance requests

### PR DESCRIPTION
**Pull request by narefyev91**
### Details
This PR adds Support for tax tracking when offline. (continue work from https://github.com/Expensify/App/pull/41379)

### Fixed Issues

$ https://github.com/Expensify/App/issues/31673
PROPOSAL:

### Tests
 - Submit Expense for Distance request
1. Go to workspace chat.
2. Start a Submit expense (money request) flow.
3. Select the distance tab
4. enter the start and stop locations then select next
5. Submit Expense

Expectation: Both the tax amount and tax rate fields cannot be modified - no right icon and the tax row is not interactive.
Edit Expense for Distance request
Go to the transaction thread and previously submitted distance expense.
Expectation: Both the tax amount and tax rate fields cannot be modified - no right icon and the tax row is not interactive.

1. Create some distance rates and taxes inside your workspace
2. Start a Submit expense (money request) flow - manual one
3. Switch tax - and verify that tax amount will be changed
4. Move to Distance tab
5. Change distance rate - and verify that tax amount is also changed

- [x] Verify that no errors appear in the JS console

### Offline tests
 - Submit Expense for Distance request
1. Go to workspace chat.
6. Start a Submit expense (money request) flow.
7. Select the distance tab
8. enter the start and stop locations then select next
9. Submit Expense

Expectation: Both the tax amount and tax rate fields cannot be modified - no right icon and the tax row is not interactive.
Edit Expense for Distance request
Go to the transaction thread and previously submitted distance expense.
Expectation: Both the tax amount and tax rate fields cannot be modified - no right icon and the tax row is not interactive.

1. Create some distance rates and taxes inside your workspace
2. Start a Submit expense (money request) flow - manual one
3. Switch tax - and verify that tax amount will be changed
4. Move to Distance tab
5. Change distance rate - and verify that tax amount is also changed

### QA Steps
 - Submit Expense for Distance request
1. Go to workspace chat.
2. Start a Submit expense (money request) flow.
3. Select the distance tab
4. enter the start and stop locations then select next
5. Submit Expense

Expectation: Both the tax amount and tax rate fields cannot be modified - no right icon and the tax row is not interactive.
Edit Expense for Distance request
Go to the transaction thread and previously submitted distance expense.
Expectation: Both the tax amount and tax rate fields cannot be modified - no right icon and the tax row is not interactive.

1. Create some distance rates and taxes inside your workspace
2. Start a Submit expense (money request) flow - manual one
3. Switch tax - and verify that tax amount will be changed
4. Move to Distance tab
5. Change distance rate - and verify that tax amount is also changed

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x]  I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/28352309/f1f98fbf-51d2-471a-919b-8975a97eb1f4

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/28352309/cc07f485-f52b-48f9-8ed3-2eda4c540227

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/28352309/06ee3c5e-48c2-4b91-b657-ffc731e5b355

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/28352309/30d941ab-7b5b-4fbe-8219-40940538bf66

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/28352309/eba3f22f-ecce-425d-8eb7-d0f7b7bccbae

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/28352309/e7b1337c-8bd4-4faa-aac9-2d24de6ae201

</details>

### CODE REVIEW

**Comment by narefyev91** [comment]
@MonilBhavsar could not make Eto's PR ready to review.
Probably we can move here.
Now we support fields visibility based on defined conditions, also new calculation from FE applied if user goes offline and distance rate is changed. I think we can start reviewing code and check if everything is good. I will prepare test cases and update videos in parallel.

**Comment by MonilBhavsar** [comment]
Thanks @narefyev91!
In QA and Tests, can we include the step to update the distance rate and ensure tax amount is updated accordingly?

File: src/libs/PolicyUtils.ts
Comment (MonilBhavsar) [inline src/libs/PolicyUtils.ts:271]: We can update to - 
```js
const isTaxEnabledForDistance = isPolicyTaxTrackingEnabled && policy?.customUnits?.[customUnitID]?.attributes?.taxEnabled;

return isDistanceRequest ? isTaxEnabledForDistance : isPolicyTaxTrackingEnabled;
```

File: src/libs/PolicyUtils.ts
Comment (MonilBhavsar) [inline src/libs/PolicyUtils.ts:272]: Thanks for update! Do we need Boolean()?

Comment (narefyev91) [reply src/libs/PolicyUtils.ts:272]: Yeah - it will break return type - because both isTaxEnabledForDistance and isPolicyTaxTrackingEnabled can be undefined

Comment (narefyev91) [reply src/libs/PolicyUtils.ts]: Done!

**Comment by MonilBhavsar** [comment]
You can do it from oldDot app. Go to Workspaces > Select a workspace > Expenses > Distance

**Review by ishpaul777 — APPROVED** [review]

File: src/libs/actions/IOU.ts
Comment (MonilBhavsar) [inline src/libs/actions/IOU.ts:429]: I think we should create separate function to update taxAmount and call it separately

File: src/pages/iou/request/step/IOURequestStepDistanceRate.tsx
Comment (MonilBhavsar) [inline src/pages/iou/request/step/IOURequestStepDistanceRate.tsx:85]: `isDistanceRequest` is redundant as this step is rendered for distance requests only.
Why `isOffline` check? We should update when user is online too.
I think we should add `isTaxTrackingEnabled()` because we want to update tax details only when it is enabled.
We should also update corresponding taxRate.
taxPercentage and taxRate to use here are - 
`var taxRate = policy?.customUnits[customUnitID].rates[customUnitRateID].attributes?.taxRateExternalID ?? '';`

Comment (MonilBhavsar) [reply src/pages/iou/request/step/IOURequestStepDistanceRate.tsx:85]: Let's add test to verify this.

When user updates distance rate, then tax rate associated with distance rate should also change

Comment (narefyev91) [reply src/pages/iou/request/step/IOURequestStepDistanceRate.tsx:85]: @MonilBhavsar why do we need to have FE calculations when user if online? It should be done on BE side

Comment (narefyev91) [reply src/pages/iou/request/step/IOURequestStepDistanceRate.tsx]: @MonilBhavsar good point - updated PR

**Comment by MonilBhavsar** [comment]
Thanks for the updates! Some more comments. Let's update this too and re test 🚀

**Comment by narefyev91** [comment]
> Thanks for the updates! Some more comments. Let's update this too and re test 🚀

done!

File: src/pages/iou/request/step/IOURequestStepDistanceRate.tsx
Comment (MonilBhavsar) [reply src/pages/iou/request/step/IOURequestStepDistanceRate.tsx:90]: This is not updated? @narefyev91

Comment (narefyev91) [reply src/pages/iou/request/step/IOURequestStepDistanceRate.tsx:90]: i think taxPercent is taken from line 59

Comment (narefyev91) [reply src/pages/iou/request/step/IOURequestStepDistanceRate.tsx:90]: Or you mean that it should be `            const taxPercentage = TransactionUtils.getTaxValue(policy, transaction, taxRateExternalID) ?? '';?
`

Comment (narefyev91) [reply src/pages/iou/request/step/IOURequestStepDistanceRate.tsx:90]: Got the point - we need to use exact tax amount based on selected taxCode from here:
`policy?.customUnits[customUnitID].rates[customUnitRateID].attributes?.taxRateExternalID ?? '';`

Comment (narefyev91) [reply src/pages/iou/request/step/IOURequestStepDistanceRate.tsx:88]: @MonilBhavsar updated!

Comment (MonilBhavsar) [reply src/pages/iou/request/step/IOURequestStepDistanceRate.tsx:88]: Thanks!

File: src/pages/iou/request/step/IOURequestStepDistanceRate.tsx
Comment (MonilBhavsar) [inline src/pages/iou/request/step/IOURequestStepDistanceRate.tsx:85]: can be const

File: src/pages/iou/request/step/IOURequestStepDistanceRate.tsx
Comment (MonilBhavsar) [inline src/pages/iou/request/step/IOURequestStepDistanceRate.tsx:88]: can be const

**Review by MonilBhavsar — COMMENTED** [review]
Thanks!
@ishpaul777 could you please retest. There are lots of updates since last review

**Comment by MonilBhavsar** [comment]
That's a feature 😄 As far as we're displaying error, it's fine UX, I think

**Review by MonilBhavsar — APPROVED** [review]

**Comment by MonilBhavsar** [comment]
Not merging because of freeze

**Comment by trjExpensify** [comment]
@MonilBhavsar what about spinning up a feature branch for these remaining XeroCon commit /app issues? Then we can merge into that once review and approved, to keep other issues moving and testing the changes in aggregate. We're doing that for the Xero integration and it's working great!

**Comment by MonilBhavsar** [comment]
Sounds good! I'll do that. I was thinking to merge this next week after things stabilizes little.

**Pull request merged**
